src/runtime/server/utils/rate-limit.ts

```diff
@@ -1,76 +1,65 @@
-import { H3Event, getRequestHeader } from 'h3'
-import { useRuntimeConfig } from '#imports'
-import type { RateLimit, RateLimitRoutes } from '../../../types'
+import {H3Event, getRequestHeader, getRequestURL} from 'h3'
+import { getRouteRules } from '#imports'
+import type { RateLimit } from '../../../types'
+import {RouteRateLimitOptions} from "../../../types";
 
 // store rate limits for each IP address and URL
 const rateLimit: RateLimit = {}
 
-// the routes we will rate limit
-const routes: RateLimitRoutes = useRuntimeConfig().public.nuxtRateLimit.routes
-
 /**
  * This function checks whether a request from a given IP address and URL should be rate limited
  *
  * If rate limited it will return the seconds until they can try again.
  *
  * @param event
  */
-export function isRateLimited(event: H3Event) {
-  try {
-    const urlWithParams = event.node.req?.url as string
-    const ip = getIP(event)
-
-    // Strip any query parameters from the URL
-    const splitURL = urlWithParams.split('?')[0]
-    // const url = urlWithParams.split('?')[0]
-
-    // configure the route settings
-    const url = routes[splitURL] ? splitURL : '/api/*'
-    const maxRequests = routes[url].maxRequests
-    const interval = routes[url].intervalSeconds * 1000
-
-    // remove any IPs & URLs that haven't been used since interval to keep object small
-    const currentTime = Date.now()
-    Object.keys(rateLimit).forEach((key: keyof RateLimit) => {
-      Object.keys(rateLimit[key]).forEach((urlKey: string) => {
-        const item = rateLimit[key][urlKey]
-        const timeSinceLastRequest = currentTime - item.firstRequestTime
-        const routeInterval = routes[urlKey].intervalSeconds * 1000
+export function getRateLimitPayload(event: H3Event) : false | { limited: boolean; limit: number; current: number; secondsUntilReset: number } {
+  const routeRules = getRouteRules(event)
+  if (!routeRules['nuxt-rate-limit']) {
+    return false
+  }
+  const { maxRequests, intervalSeconds, route } = routeRules['nuxt-rate-limit'] as RouteRateLimitOptions
+  const intervalMs = intervalSeconds * 1000
+  const ip = getIP(event)
```

Comment:
I made a PR to implement this in h3, you may be able to use this instead if it's merged https://github.com/unjs/h3/pull/503 (credits to you for the code :+1: )